static-site/src/components/ListResources/types.ts

```diff
@@ -31,6 +49,36 @@ export interface Resource {
   updateCadence?: UpdateCadence
 }
 
+export interface DisplayNamedResource extends Resource {
+  displayName: ResourceDisplayName
+}
+
+export interface VersionedResource extends Resource {
+  lastUpdated: string  // date
+  firstUpdated: string  // date
+  dates: string[]
+  nVersions: number
+  updateCadence: UpdateCadence
+}
+
+export function convertVersionedResource(resource: Resource): VersionedResource {
+  if (resource.lastUpdated !== undefined &&
+      resource.firstUpdated !== undefined &&
+      resource.dates !== undefined &&
+      resource.nVersions !== undefined &&
+      resource.updateCadence !== undefined) {
+    return {
+      ...resource,
+      lastUpdated: resource.lastUpdated,
+      firstUpdated: resource.firstUpdated,
+      dates: resource.dates,
+      nVersions: resource.nVersions,
+      updateCadence: resource.updateCadence
+    }
```

Comment:
I don't understand what this code is doing — you just checked that all those properties exist on `resource`, so they will be there in the spread version, and then you're overwriting them with ...the same values? 

How is this different from just doing `return resource;`?

Reply:
`return resource` gives a type error:

```
error TS2322: Type 'Resource' is not assignable to type 'VersionedResource'.
  Types of property 'lastUpdated' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.
```

Another alternative is `return resource as VersionedResource` but that loses the tight coupling between the condition and return value.

Reply:
This was the main TS concept I dove into (see comments in [this issue](https://github.com/nextstrain/auspice/issues/1889)) and as best as I can tell changing the type of an object (i.e. the same reference) goes against the grain of TS. Most suggestions involved using something like Victor's done to change the type. (You can cast - it's not like TS forbids it - but it seems an antipattern.)

Reply:
My impression is also that we should avoid casting (and similarly non-null assertions) if possible, even if the proper way is more verbose like in this example. I'll explore removing them and enforcing that in a separate PR.

Reply:
> return resource gives a type error:

What about `return {...resource}` - is TS smart enough to match the non-undefined checks and therefore allow the newly created object to be of `VersionedResource` type?

Reply:
No, this is the error:

```
Type '{ name: string; displayName?: ResourceDisplayName; groupName: string; nameParts: string[]; sortingName: string; url: string; lastUpdated?: string; firstUpdated?: string; dates?: string[]; nVersions?: number; updateCadence?: UpdateCadence; }' is not assignable to type 'VersionedResource'.
  Types of property 'lastUpdated' are incompatible.
    Type 'string | undefined' is not assignable to type 'string'.
      Type 'undefined' is not assignable to type 'string'.
```

The first object shape is just a reflection of `resource: Resource` from the function signature. `resource: Resource | VersionedResource` doesn't work either.

Reply:
Separate PR for stricter types: #1077

Reply:
> The first object shape is just a reflection of `resource: Resource` from the function signature. `resource: Resource | VersionedResource` doesn't work either.

Maybe only marginally more attractive, but I wonder if turning this into something like a wrapped type guard would work? Something like 

```typescript
function foo(bar: Resource): VersionedResource {
  const _guard = (x:Resource): x is VersionedResource => 
    return x.lastUpdaded && the other checks here;

  if (_guard(bar)) { return bar }
  else { throw }
}
```

Reply:
Hmm, I find that less attractive because `x is VersionedResource` is effectively a type assertion. A type error won't be raised if you accidentally forget to check one of the the properties in the return statement. So it's about the same as `return resource as VersionedResource`.

Reply:
> Hmm, I find that less attractive because `x is VersionedResource` is effectively a type assertion.

...a type assertion supported by exactly the same logic that underlies the spread + override `return` in your code.

> A type error won't be raised if you accidentally forget to check one of the the properties in the return statement.

I mean, yes, if you write a bug in the code, the code is going to be buggy. :grin:

> So it's about the same as `return resource as VersionedResource`.

I disagree; there's zero validation associated with a cast on the return.

Reply:
> if you write a bug in the code, the code is going to be buggy

Yes, and we should leverage TypeScript to catch the bug at compile time. I haven't articulated my point very well. This example bug + screenshots might help.

Scenario: a property `newVersionedProperty` is added. Just like `lastUpdated`, `firstUpdated`, etc., it is optional on `Resource` and required on `VersionedResource`. The bug is that we forget to add it to this function.

![typing](https://github.com/user-attachments/assets/ae4c1d54-6efe-4b95-9b3a-94430a63d4bd)

1. This is the code in the PR. There is TypeScript error: `Types of property 'newVersionedProperty' are incompatible`. Useful!
2. This is `return resource as VersionedResource`. No TypeScript error.
3. This is the guard `resource is VersionedResource`. No TypeScript error.